compactor2/src/components/compact/query_chunk.rs

```diff
@@ -0,0 +1,278 @@
+        //      each other and overlapped L1 files) first.
+        //    . Level-N+1 files are results of compacting Level-N and/or Level-N+1 files, their `created_at`
+        //      can be after the `created_at` of other Level-N files but they may include data loaded before
+        //      the other Level-N files. Hence we should never use `created_at` of Level-N+1 files to order
+        //      them with Level-N files.
+        //    . We can only compact different sets of files of the same partition concurrently into the same target_level.
+        // We can use the following rules to set order of the chunk of its (compaction_level, target_level) as follows:
+        //    . compaction_level < target_level : the order is `created_at`
+        //    . compaction_level == target_level : order is `0` to make sure it is in the front of the ordered list.
+        //      This means that the chunk of `compaction_level == target_level` will be in arbitrary order and will be
+        //      fine as long as they are in front of the chunks of `compaction_level < target_level`
+
+        match (self.compaction_level, self.target_level) {
+            (CompactionLevel::Initial, CompactionLevel::FileNonOverlapped)
+            | (CompactionLevel::FileNonOverlapped, CompactionLevel::Final) => {
+                ChunkOrder::new(self.created_at().get())
+            }
+            (CompactionLevel::FileNonOverlapped, CompactionLevel::FileNonOverlapped)
+            | (CompactionLevel::Final, CompactionLevel::Final) => ChunkOrder::new(0),
+            _ => {
+                panic!(
+                    "Invalid compaction level combination: ({:?}, {:?})",
+                    self.compaction_level, self.target_level
+                );
+            }
+        }
+    }
+
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+}
+
+/// Convert to a QueryableParquetChunk
+pub fn to_queryable_parquet_chunk(
+    file: ParquetFile,
+    store: ParquetStorage,
+    table_schema: &TableSchema,
+    partition_sort_key: Option<SortKey>,
+    target_level: CompactionLevel,
+) -> QueryableParquetChunk {
+    let column_id_lookup = table_schema.column_id_map();
+    let selection: Vec<_> = file
+        .column_set
+        .iter()
+        .flat_map(|id| column_id_lookup.get(id).copied())
+        .collect();
+    let table_schema: Schema = table_schema
+        .clone()
+        .try_into()
+        .expect("table schema is broken");
+    let schema = table_schema
+        .select_by_names(&selection)
+        .expect("schema in-sync");
+    let pk = schema.primary_key();
+    let sort_key = partition_sort_key
+        .as_ref()
+        .map(|sk| sk.filter_to(&pk, file.partition_id.get()));
+
+    let partition_id = file.partition_id;
+    let min_time = file.min_time;
+    let max_time = file.max_time;
+    let compaction_level = file.compaction_level;
+
+    trace!(
+        parquet_file_id=?file.id,
```

Comment:
This applies to all `debug!`/`trace!`/`info!` calls: debug-printing IDs is very verbose and makes them impossible to parse later. You end up with logs saying `parquet_file_id=ParquetFileId(123)`. You can avoid this by passing the contained integer directly to the tracing library:

```suggestion
        parquet_file_id=file.id.get(),
```

Note that there is NOT `?` or `%` there, because the lib understands primitive types like integers out-of-the-box. This also means that any logging sink will get proper integers, which helps with machine-parsing.